File: Two_Pointers/C++/11_container_with_most_water.cpp
/*
* RUNTIME: 66 ms - Beats 75.96% of submissions
* MEMORY - 61.50 MB - Beats 5.99% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
using namespace std;


/* Thoughtprocess:
* 1. initailize a left and right pointer
* 2. Comapare the maximum area with the current area
* 3. If left height is less than right height, move the left pointer to the right
* 4. If right height is less than left height, move the right pointer to the left
* 5. Repeat untill no possibilities are left
* 6. Return the maximum area
* 
*/


class Solution {
public:
    int maxArea(vector<int>& height) {
        int left = 0;
        int right = height.size() - 1;
        int area = 0;

        while (left < right)
        {
            area = max(area, min(height[left], height[right]) * (right - left));
            if (height[left] < height[right])
                left++;
            else
                right--;
        }
        return area;
    }
};

File: Two_Pointers/Python/11_container_with_most_water.py
# RUNTIME: 532 ms, faster than 78.47% of python3 Submissions. 
# MEMORY USAGE: 29.47 MB, less than 63.59% of python3 Submissions.

from __future__ import annotations
from ast import *


# My thougtprocess:
# 1. Define left and right pointers
# 2. check the maximum area between left and right and update maxArea
# 3. if sum of left height is smaller than right height, we need to increase left
# 4. return the max area
height = [1,8,6,2,5,4,8,3,7]

class Solution:
    def maxArea(self, height: List[int]) -> int:
        left = 0
        right = len(height) - 1
        maxArea = 0

        while left < right:
            maxArea = max(maxArea, min(height[left], height[right]) * (right - left))

            if height[left] < height[right]:
                left += 1
            else:
                right -= 1

        return maxArea